### PR TITLE
[FLINK-32355][table] Support to list procedures

### PR DESCRIPTION
**Pull request by luoyuxia**
## What is the purpose of the change
To make Flink support list procedures.

## Brief change log
  -  Add an interface in Catalog to list procedure
  -  Add a parsing rule for `Show Procedure` to parse the syntax to `SqlShowProcedures` in flink-sql-parser
  -  Add a converter to convert `SqlShowProcedures`  to `ShowProceduresOperation`.
  -  Add the actual execution logic in `ShowProceduresOperation`.

## Verifying this change
UT&IT

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

**Review by chucheng92 — COMMENTED** [review]
Shall we add show procedures in user docs?

**Comment by chucheng92** [comment]
Most of the logic is the same, but there are some common logics, can you review it for me, and I'm sorry for the late reply.

**Comment by luoyuxia** [comment]
> Shall we add show procedures in user docs?

I would like to do it in another pr to keep this pr fouces.

**Comment by luoyuxia** [comment]
> @luoyuxia Hi, yuxia. I also submit a PR #22903. Most of the logic is the same, but there are some common logics, can you help to review it? thanks and i'm sorry for the late reply.

Review done. PTAL.

**Comment by chucheng92** [comment]
Hi, it looks good to me. but ci failed, maybe error usage of '<>' in javadoc.

**Review by chucheng92 — APPROVED** [review]
LGTM, +1

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlShowProcedures.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    private final boolean isILike;
+
+    private final SqlCharStringLiteral likeLiteral;
+
+    public SqlShowProcedures(
+            SqlParserPos pos,
+            String preposition,
+            SqlIdentifier databaseName,
+            boolean notLike,
+            boolean isILike,
```
[the remark this answers is not in the document]
Comment (luoyuxia) [reply flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlShowProcedures.java]: done.

File: flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl [hunk 2]
```diff
@@ -455,6 +455,46 @@ SqlShowFunctions SqlShowFunctions() :
...
+/**
+* Parses a show functions statement.
+* SHOW PROCEDURES [ ( FROM | IN ) [catalog_name.]database_name ] [ [NOT] (LIKE | ILIKE) pattern;
+*/
+SqlShowProcedures SqlShowProcedures() :
+{
+    String prep = null;
+    SqlIdentifier databaseName = null;
+    boolean notLike = false;
+    boolean isIlike = false;
```
[the remark this answers is not in the document]
Comment (luoyuxia) [reply flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl]: done.

**Comment by chucheng92** [comment]
@luoyuxia hi, yuxia. I leave a few minor comments, PTAL.

**Comment by luoyuxia** [comment]
@luoyuxia Thanks for your review. I hope I have addressed your comments.

**Comment by chucheng92** [comment]
> @luoyuxia Thanks for your review. I hope I have addressed your comments.

Hi, yuxia. i mean your likeType and likePattern both @Nullable. If user give a not null likeType and null likePattern, can we throw exception in Constructor phase rather than rely on method implementation.

File: flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl (hunk 2 above)
[the remark this answers is not in the document]
Comment (luoyuxia) [reply flink-table/flink-sql-parser/src/main/codegen/includes/parserImpls.ftl]: I would like to let it to be a string, and then convert to `LikeTypeEnum` after sqlnode conversion phase.

**Comment by luoyuxia** [comment]
@chucheng92 Thanks. I got you. I have updated it.

**Comment by chucheng92** [comment]
Looks good. +1 for the PR. @luoyuxia

**Review by chucheng92 — APPROVED** [review]

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlShowProcedures.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    // different like type such as like, ilike
+    @Nullable private final String likeType;
+    @Nullable private final SqlCharStringLiteral likeLiteral;
+
+    public SqlShowProcedures(
+            SqlParserPos pos,
+            @Nullable String preposition,
+            @Nullable SqlIdentifier databaseName,
+            boolean notLike,
+            @Nullable String likeType,
```
[the remark this answers is not in the document]
Comment (luoyuxia) [reply flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlShowProcedures.java:64]: fixed

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ShowProceduresOperation.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+            @Nullable String catalogName,
+            @Nullable String databaseName,
+            boolean notLike,
+            @Nullable String likeType,
+            @Nullable String sqlLikePattern) {
+        this.preposition = preposition;
+        this.catalogName = catalogName;
+        this.databaseName = databaseName;
+        this.notLike = notLike;
+        this.likeType = likeType == null ? null : LikeType.of(likeType);
```
[the remark this answers is not in the document]
Comment (luoyuxia) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ShowProceduresOperation.java]: fixed

**Comment by luoyuxia** [comment]
@chucheng92 Thanks for the review. WIll merge later.

**Pull request merged**